lib/mayaUsd/ufe/UsdAttributes.cpp

```diff
@@ -341,5 +365,149 @@ UsdAttributes::getUfeTypeForAttribute(const PXR_NS::UsdAttribute& usdAttr) const
     return Ufe::Attribute::kInvalid;
 }
 
+#if (UFE_PREVIEW_VERSION_NUM >= 4024)
+// Helpers for validation and execution:
+bool UsdAttributes::canAddAttribute(
+    const UsdSceneItem::Ptr&    item,
+    const std::string&          name,
+    const Ufe::Attribute::Type& type)
+{
+    // See if we can edit this attribute, and that it is not already part of the schema or node
+    // definition
+    if (!item || !item->prim().IsActive() || UsdAttributes(item).hasAttribute(name)) {
+        return false;
+    }
+
+    // Since we can always fallback to adding a custom attribute on any UsdPrim, we accept.
+    return true;
+}
+
+Ufe::Attribute::Ptr UsdAttributes::doAddAttribute(
+    const UsdSceneItem::Ptr&    item,
+    const std::string&          name,
+    const Ufe::Attribute::Type& type)
+{
+    // See if we can edit this attribute, and that it is not already part of the schema or node
+    // definition
+    if (!item || !item->prim().IsActive() || UsdAttributes(item).hasAttribute(name)) {
```

Comment:
Why not just call canAddAttribute and return nullptr if it returns false to avoid the code duplication?